R/slide.R

```diff
@@ -363,3 +364,403 @@ epi_slide <- function(x, f, ..., before, after, ref_time_values,
+    # should be used instead.
+    if (anyDuplicated(.data_group$time_value) > 0) {
+      cli_abort(c(
+          "group contains duplicate time values. Using `epi_slide_mean` on this
+            group will result in incorrect results",
+          "i" = "Please change the grouping structure of the input data so that
+            each group has non-duplicate time values",
+          "i" = "Use `epi_slide` to aggregate across groups"
+        ),
+        class = "epiprocess__epi_slide_mean__duplicate_time_values",
+        epiprocess__data_group = .data_group,
+        epiprocess__group_key = .group_key
+      )
+    }
+
+    roll_output <- data.table::frollmean(
+      x = .data_group[, col_name], n = m, align = "right", ...
+    )
+
+    if (after >= 1) {
+      # Right-aligned `frollmean` results' `ref_time_value`s will be `after`
+      # timesteps ahead of where they should be. Shift results to the left by
+      # `after` timesteps.
+      .data_group[, result_col_name] <- purrr::map(roll_output, function(.x) {
+          c(.x[(after + 1L):length(.x)], rep(NA, after))
+        }
+      )
+    } else {
+      .data_group[, result_col_name] <- roll_output
+    }
+
+    return(.data_group)
+  }
+
+  result <- mutate(x, .real = TRUE) %>%
+    group_modify(slide_one_grp, ..., .keep = FALSE)
+
+  result <- result[result$.real, ]
+  result$.real <- NULL
+
+  if (all_rows) {
+    result[!(result$time_value %in% ref_time_values), result_col_name] <- NA
+  } else if (user_provided_rtvs) {
+    result <- result[result$time_value %in% ref_time_values, ]
+  }
+
+  if (as_list_col) {
+    result[, result_col_name] <- purrr::map(result_col_name,
+      function(.x) {
+         tmp <- result[[.x]]
+         tmp[is.na(tmp)] <- list(NULL)
+         as.list(tmp)
+      }
+    )
+  }
+
+  if (!is_epi_df(result)) {
+    # `all_rows` and `as_list_col` handling strip epi_df format and metadata.
+    # Restore them.
+    result <- reclass(result, attributes(x)$metadata)
+  }
+
+  return(result)
+}
+
+#' Make a complete date sequence between min(x$time_value) and max
+#' (x$time_value). Produce lists of dates before min(x$time_value) and after
+#' max(x$time_value) for padding initial and final windows to size `n`.
+#'
+#' @importFrom checkmate assert_function
+#' @noRd
+full_date_seq <- function(x, before, after, time_step) {
+  pad_early_dates <- c()
+  pad_late_dates <- c()
+
+  # If dates are one of tsibble-provided classes, can step by numeric. `tsibble`
+  # defines a step of 1 to automatically be the quantum (smallest resolvable
+  # unit) of the date class. For example, one step = 1 quarter for `yearquarter`.
+  #
+  # `tsibble` classes apparently can't be added to in different units, so even
+  # if `time_step` is provided by the user, use a unit step.
+  if (inherits(x$time_value, c("yearquarter", "yearweek", "yearmonth")) ||
+    is.numeric(x$time_value)) {
+    all_dates <- seq(min(x$time_value), max(x$time_value), by = 1L)
```

Comment:
I'm not clear on what your desired behavior is. Can you give an example of "an `epix_slide()` with an optimistic `ref_time_values`"?

Right now, neither `epi_slide[_mean]` checks number of obs in the input data. If we want to raise an error for 0-row input data, it's probably clearer if we do that in `epi_slide[_mean]` rather than in the `full_date_seq` helper.

A 1-row input can cause a warning (or error, I suppose) that says the user should explicitly set `time_type`.

Reply:
> Can you give an example of "an epix_slide() with an optimistic ref_time_values"?

By "optimistic", I meant an `epix_slide()` given a `ref_time_values` that includes something less than the max version.  This could be from trying to line up with another data source, or accidentally trying to set `ref_time_values` based on the `time_value` range rather than `version` range; here's an example of the latter + the error we get:

```{r}
test_date_range <- range(archive_cases_dv_subset$DT$time_value)
test_dates <- seq(test_date_range[[1L]], test_date_range[[2L]], by = 7L)
archive_cases_dv_subset %>%
  epix_slide(
    before = 6L, ref_time_values = test_dates,
    ~ .x %>% group_by(geo_value) %>% epi_slide_mean("percent_cli", before = 6)
  )
#> Error in seq.int(0, to0 - from, by) : 'to' must be a finite number
#> In addition: Warning messages:
#> 1: In min.default(numeric(0), na.rm = FALSE) :
#>   no non-missing arguments to min; returning Inf
#> 2: In max.default(numeric(0), na.rm = FALSE) :
#>   no non-missing arguments to max; returning -Inf
```

There's also an `epix_slide()` component of improving the error message above, but the part that matters here is just the inner computation.

```{r}
jhu_csse_daily_subset %>%
  filter(time_value <= as.Date("2020-01-01")) %>%
  group_by(geo_value) %>%
  epi_slide_mean("cases", before = 6L)
#> Error in seq.int(0, to0 - from, by) : 'to' must be a finite number
#> In addition: Warning messages:
#> 1: In min.default(numeric(0), na.rm = FALSE) :
#>   no non-missing arguments to min; returning Inf
#> 2: In max.default(numeric(0), na.rm = FALSE) :
#>   no non-missing arguments to max; returning -Inf
```
It'd be nice to have some better feedback here, explicitly telling us there's an issue with 0-row inputs.

> If we want to raise an error for 0-row input data, it's probably clearer if we do that in epi_slide[_mean] rather than in the full_date_seq helper.

Agreed, that'd improve the "Error in" line.

Looks like we also have this issue with `epi_slide()` + `mean()` if not using `as_list_col = TRUE`, so maybe people haven't actually run into it yet?
```r
jhu_csse_daily_subset %>%
  filter(time_value <= as.Date("2020-01-01")) %>%
  group_by(geo_value) %>%
  epi_slide(before = 6L, cases_7dav = mean(.x$cases))
#> Error in `.row_names_info()` at dplyr/R/utils.R:18:3:
#> ! Can't subset columns that don't exist.
#> ✖ Column `cases_7dav` doesn't exist.
#> Run `rlang::last_trace()` to see where the error occurred.
> jhu_csse_daily_subset %>%
  filter(time_value <= as.Date("2020-01-01")) %>%
  group_by(geo_value) %>%
  epi_slide(before = 6L, cases_7dav = mean(.x$cases), as_list_col = TRUE)
#> An `epi_df` object, 0 x 7 with metadata:
#> * geo_type  = state
#> * time_type = day
#> * as_of     = 2024-01-26 09:27:32
#> 
#> # A tibble: 0 × 7
#> # Groups:   geo_value [0]
#> # ℹ 7 variables: geo_value <chr>, time_value <date>, cases <dbl>, cases_7d_av <dbl>,
#> #   case_rate_7d_av <dbl>, death_rate_7d_av <dbl>, cases_7dav <list>
```

> A 1-row input can cause a warning (or error, I suppose) that says the user should explicitly set time_type.

(I'm hoping no extra handling's needed here for `epi_slide_mean()` and this is just something we might want to do for `new_epi_df()` sometime.)

Reply:
> Looks like we also have [an unclear error message in 0-row case] with epi_slide() + mean() if not using as_list_col = TRUE, so maybe people haven't actually run into it yet?

Will leave this particular issue as-is for now, given planned changes to `epi_slide`.

Reply:
Addressed the 0-row error in `epi_slide_mean` in https://github.com/cmu-delphi/epiprocess/pull/400/commits/bf5f8dfa8bc92a0819a48f7f4854345fe940e8ef